kafka-rest/src/main/java/io/confluent/kafkarest/response/StreamingResponse.java

```diff
@@ -141,31 +154,56 @@ private ResultOrError handleNext(T result, @Nullable Throwable error) {
   private static class InputStreamingResponse<T> extends StreamingResponse<T> {
 
     private final MappingIterator<T> mappingIteratorInput;
+    private final StreamingResponseIdleTimeLimiter idleTimeLimiter;
+
+    private Throwable hasNextError = null;
+    private boolean closed = false;
 
     private InputStreamingResponse(
-        MappingIterator<T> mappingIteratorInput, ChunkedOutputFactory chunkedOutputFactory) {
+        MappingIterator<T> mappingIteratorInput,
+        ChunkedOutputFactory chunkedOutputFactory,
+        StreamingResponseIdleTimeLimiter idleTimeLimiter) {
       super(chunkedOutputFactory);
       this.mappingIteratorInput = requireNonNull(mappingIteratorInput);
+      this.idleTimeLimiter = requireNonNull(idleTimeLimiter);
     }
 
     public void close() {
-      try {
-        mappingIteratorInput.close();
-      } catch (IOException e) {
-        log.error("Error when closing the request stream.", e);
-      }
+      closed = true;
     }
```

Comment:
Why has the close of the mapping iterator been removed here?  If I remember correctly, if the mapping Iterator isn't closed off, then it continues to read data from the socket, even after the sink is closed.

Reply:
You remember almost correctly, if I remember correctly :). The error that you were seeing is that there was no `!responseQueue.isClosed()` in the `while (hasNext() && !responseQueue.isClosed())` loop. Therefore the while loop was not exiting when you closed the sink before exhausting the source. We don't actually have to close the input byte-stream. That is taken care of by jetty in the onComplete of the HTTP request. We just need this while-loop to exit, which it will. The reason a removed the early close of the input byte-stream is that, in some cases, if we close the input byte-stream before flushing the output byte-stream, this will cause a pipe broken in the client, and it will refuse to read anything we have written back (that it hasn't read yet).